### ISSUE TITLE
`reactive.event` is not caching properly

### ISSUE THREAD

**Issue by schloerke** [rstudio/py-shiny#691]
When reading a reactive value that has been decorated by a `reactive.event`, if the value is read again at a later date, the original reactive is recalculated. 

I would expect the existing value to be recycled.

Reprex app: https://shinylive.io/py/editor/#code=NobwRAdghgtgpmAXGKAHVA6VBPMAaMAYwHsIAXOcpMAMwCdiYACAZwAsBLCbJjmVYnTJMAgujxM6lACZw6EgK4cJUqITIcAbnBVwAjgB0IfAUKbSoFDfCNG0qAPpKmAXiZKsUAOZwHNADZK0gAURkzh7hwYMFB0ANbSxADuEKEQERlMBmDZuemZ4QDEogpkjJZw0pJwqFIAHohhBVn5BQBUTADC-hyEcUwARqVlECxMAAYA8hBw40xQEFXjACpJxONN7V09fYPDpBOr60zB3lBcAJSbmdcZAHLEGoRwTGzJTGRsL3B1qHDqlSYxFKqFKE1Is3MxDgYzabQgjzh7lQFgoGCYy04Yw4YxIdCk6kGcDYUE0HEEGFuESp4QAEsk4Np5JIFolmOMIXMFCiKix0ZicbwxgoZr9-hQqgNiaTyXRKa0bjkldcLnhrh4uKCyA41BpSA4hmQRqEwBDshJstM4NlVeqoprSjr1OSIAb9qlsmQ1uasmAjja1QqPMCyFqHBQ6tqmQNLHwTWRIwG7RgyN4+SxCHQOKgyGlmr7cjkFRkACTBRKEBTwcgXDCkYIAcnYXGwiBIEBmAOkDYkNBFzvrjMoZAuTBANMyFarw4wegUcmwAGU4P5xYITYUzWBa4QdnFghcANwTjJT6tkWfzuhLldrugbr3EG0YXe9fdHk8RFhwMjLPhwENgj7CAB1SUdx2LfMIjPGc5wXZdV3UddskKR9n1fPoD2PSCCgAXwkABWAAGIirhwphcI-cjCzyDIyLI2xFjgGhWDkJlggdMgJBDLUJG-FgWBdC5Ggga4AAFVGdbQMCHcgOIgLU6xmMjT2YllFkYBwIQPET81qLhc2yCEAEIbQnKQyAUOh0hoIyZkQMdUR-f8MCc6w4AwBEkgPXC8nEySNGk2Tc04lM1hU6C1LoVlNMfHSJ30uTPTWUzt3Mn8rJs7IQGijSYC0mYfIMIxHwckA3Jcir4E85IiqLfy4F1LQPM6KB-EIa5ZBYn4-i7Aq4Hi8ipD0eTFO0i4IoKJIOE+aomuknFiH8CpBqgiJEsMsAevFSoHJMszyIyCzMqYWzTXsxyKnc1yrpcry6sYjIJMaqSWrajqFS6phtr6uLhPSkbQr+ybMmm2aAuajBFuWihVrWpgNvjFKDvh6pLOs07sp+iV+qKkq1jKqqPKJmrvIuXyi3zcSeNKBqmLlCMyE6tSE1zf6hoyjGzoAJRewKXgAUW0cgTkrfFhwkGNpH+owcpi-K-t82wjAFsUuxOLxiGIaXGll7HKnDNYHtEk37FcUR0FOdAnGUVi6CZMiwFwgBdIA

![Screenshot 2023-08-21 at 11 53 33 AM](https://github.com/rstudio/py-shiny/assets/93231/dbab97e8-5509-49aa-a75d-193e7edfb1cb)

**Comment by wch** [comment]
I believe the problem is due to a missing `reactive.Calc`.

For future reference, here is the code for the example app:

<details>
<summary>App code</summary>

```python
from shiny import App, render, ui, reactive, req
import datetime

app_ui = ui.page_fluid(
    ui.markdown(
        """
        # Automated reprex:
        
        * Click buttons `One` and `Two`
        * Click button `Two` (again)
        
        Notice how the expected output `one` does **not** update. This is correct behavior.
        
        However, random `one` updates. This is unexpected behavior.
        """
    ),
    ui.input_action_button("one", "One"),
    ui.input_action_button("two", "Two"),
    ui.output_text_verbatim("txt"),
    ui.tags.script(
        """
        $(document).on('shiny:connected', function(event) {
            document.querySelector("#one").click();
            document.querySelector("#two").click();
            setTimeout(function() {
                document.querySelector("#two").click();
            }, 500)
        });
        """
    )
)


def server(input, output, session):

    @reactive.event(input.one)
    def random_one():
        print("one!")
        return f"one: {datetime.datetime.now()}"

    @reactive.event(input.two)
    def random_two():
        print("two!")
        return f"{random_one()}\ntwo: {datetime.datetime.now()}"

    @reactive.Calc
    def expected_one():
        req(input.one())
        with reactive.isolate():
            print("expected: one!")
            return f"one: {datetime.datetime.now()}"

    @reactive.Calc
    def expected_two():
        req(input.two())
        with reactive.isolate():
            print("two!")
            return f"{expected_one()}\ntwo: {datetime.datetime.now()}"
        
    @output
    @render.text
    def txt():
        return f"Reactive Event (current, bad):\n{random_two()}\n\nExpected (good):\n{expected_two()}"


app = App(app_ui, server)
```
</details>



I noticed that `random_one()` and `random_two()` don't have a `reactive.Calc` on them:

```python
    @reactive.event(input.one)
    def random_one():
        print("one!")
        return f"one: {datetime.datetime.now()}"

    @reactive.event(input.two)
    def random_two():
        print("two!")
        return f"{random_one()}\ntwo: {datetime.datetime.now()}"
```

If you add `reactive.Calc`, then the app works as expected:

```python
    @reactive.Calc
    @reactive.event(input.one)
    def random_one():
        print("one!")
        return f"one: {datetime.datetime.now()}"

    @reactive.Calc
    @reactive.event(input.two)
    def random_two():
        print("two!")
        return f"{random_one()}\ntwo: {datetime.datetime.now()}"
```

The `reactive.Calc` is what provides caching; without it, no caching occurs, and so when `random_two` re-executes, it causes `random_one` to re-execute (without caching).

In summary, it's a bug in the app code. It would be nice if we could alert users to this kind of bug, but I'm not sure offhand if there's a way that we can do that without causing other problems.

**Comment by schloerke** [comment]
Yup. I can see the fix working on my side. Thank you!


---------------------------------------

Do you think it'd be something that we return a function that will "always raise a runtime error"? But if the function is absorbed by a render or calc method, then the "always raise a runtime error" is removed?

Ex: Similar to CardItem that wraps an object and is unwrapped at runtime, we could wrap the method in a callable class that always throws with a nice error message, but Calc or render methods could _unwrap_ the class.

**Comment by wch** [comment]
Hm, I'm not sure that it's a good idea to restrict a function decorated with `@reactive.event` to _only_ be consumable by `@reactive.Calc`, `@reactive.Effect`,  or`@render.xx`, although I do think that would be by far the common case.

I can think of a few reasons: 
- Various decorators can be used in a stack, and someone might want to put other decorators on top of `@reactive.event`.
- In a more general sense, `@reactive.event` basically means that a function depends on specific objects -- imposing this new requirement to make it only consumable by specific other things (like `@reactive.Calc`).
